src/ngx_http_lua_misc.c

```diff
@@ -286,6 +286,34 @@ ngx_http_lua_ffi_headers_sent(ngx_http_request_t *r)
 
     return r->header_sent ? 1 : 0;
 }
+
+
+int
+ngx_http_lua_ffi_get_conf_env(u_char *name, u_char **value_buf,
+    size_t *value_len)
+{
+
+    ngx_uint_t            i;
+    ngx_str_t            *var;
+    ngx_core_conf_t      *ccf;
+
+    ccf = (ngx_core_conf_t *) ngx_get_conf(ngx_cycle->conf_ctx,
```

Comment:
Mhhhh that's not the results I experienced with my tests. Both the signaller and the master process show that they have the new values upon a reload. Same with a binary upgrade.

Reply:
@thibaultcha 
I tried it with gdb, and they looks like not the same.
```
(gdb) b ngx_http_lua_init
...
(gdb) p (ngx_core_conf_t*)ngx_cycle->conf_ctx[ngx_core_module.index]
$10 = (ngx_core_conf_t *) 0x12927d0
(gdb) p *((ngx_str_t *)$10->env->elts)
$11 = {
  len = 3, 
  data = 0x1294520 "VAR=old"
}
(gdb) p (ngx_core_conf_t*)cf->cycle->conf_ctx[ngx_core_module.index]
$12 = (ngx_core_conf_t *) 0x12718b0
(gdb) p *((ngx_str_t *)$12->env->elts)
$13 = {
  len = 3, 
  data = 0x12bd110 "VAR=new"
}
(gdb) b ngx_http_lua_init_globals
...
(gdb) p (ngx_core_conf_t*)cycle->conf_ctx[ngx_core_module.index]
$14 = (ngx_core_conf_t *) 0x12718b0
(gdb) p *((ngx_str_t *)$14->env->elts)
$15 = {
  len = 3, 
  data = 0x12bd110 "VAR=new"
}
```

Reply:
I run it with
```nginx
master on;
daemon off;
worker_processes  1;
#env VAR=old;
env VAR=new;
```

Reply:
> master on;

Do you mean `master_process on;`?

I made the following test:
```
master_process on;
daemon off;

events {}

env foo=old;
#env foo=new;

http {
    init_by_lua_block {
        require "resty.core"
        process = require "ngx.process"

        cmd = string.format([[
            echo '[%d %s] foo=%s' >> env.log
        ]], ngx.worker.pid(), process.type(), tostring(os.getenv("foo")))

        os.execute(cmd)
    }
}
```

Then running:
```
$ nginx -p .
```

I observe:
```
$ cat env.log
[25672 single] foo=old
```

Then, I replace the `env` directive in the test configuration:
```
#env foo=old;
env foo=new;
```

And run:
```
$ kill -HUP `cat logs/nginx.pid`
$ cat env.log
[25672 single] foo=old
[25672 master] foo=new
```

So in this test, the value is being updated.

I am running a development build of ngx_lua with this branch, no OpenResty-maintained patches. Could that be an issue and explain a difference in behaviour? I will give it a try if you still do not observe the same results as I do.

Reply:
@spacewander I just verified with gdb myself; here is the reason why: https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_module.c#L763

Reply:
This probably means that this feature should be tested form all code paths that can call `init_handler` (with and without `lua_shared_dict`), and updates to the `env` directives would obviously not work with `lua_code_cache off;` when the `init_by_lua*` directives are invoked while initializing the Lua VM (https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_util.h#L296).

Reply:
@thibaultcha 
Thank you for pointing it out! It is my first time to notice this trick. The trick seems to be used to simpify the implementation of semaphore.

For `lua_code_cache off;`, since we don't re-parse the Nginx configuration, updates to the `env` directives won't work is expected.